vision/spatial_transformer/spatial_transformer.jl

```diff
@@ -0,0 +1,200 @@
+# Spatial Transformer Network
+
+#=
+In this tutorial we'll build a spatial transformer network that will transform MNIST
+digits for classification by a convolutional network
+
+* [Spatial Transformer Networks](https://proceedings.neurips.cc/paper/2015/hash/33ceb07bf4eeb3da587e268d663aba1a-Abstract.html)
+
+=#
+
+using LinearAlgebra, Statistics
+using Flux, CUDA, Zygote
+using Flux: batch, onehotbatch, flatten, unsqueeze
+using MLDatasets
+using IterTools: partition
+using Plots
+using ProgressMeter
+using ProgressMeter: Progress
+## =====
+
+args = Dict(
+    :bsz => 64, # batch size
+    :img_size => (28, 28), # mnist image size
+    :n_epochs => 40, # no. epochs to train
+)
+
+
+## ==== GPU
+dev = has_cuda() ? gpu : cpu
+
+## ==== Data functions
+"split your data into batches of size bsz"
+get_batches(x; bsz=args[:bsz]) = batch.(partition(x, bsz))
+
+function get_mnist_data()
+    train_digits, train_labels = MNIST(split=:train)[:]
+    test_digits, test_labels = MNIST(split=:test)[:]
+    xs_train, xs_test = map(get_batches, (eachslice(train_digits, dims=3), eachslice(test_digits, dims=3)))
+
+    ys_train = map(x -> Float32.(onehotbatch(x, 0:9)), get_batches(train_labels))
+    ys_test = map(x -> Float32.(onehotbatch(x, 0:9)), get_batches(test_labels))
+
+    return xs_train, xs_test, ys_train, ys_test
+end
+
+## ==== interpolation functions
+
+"generate sampling grid 3 x (width x height) x (batch size)"
+function get_sampling_grid(width, height; args=args)
+    x = collect(LinRange(-1, 1, width))
+    y = collect(LinRange(-1, 1, height))
+    x_t_flat = reshape(repeat(x, height), 1, height * width)
+    y_t_flat = reshape(repeat(transpose(y), width), 1, height * width)
+    all_ones = ones(eltype(x_t_flat), 1, size(x_t_flat)[2])
+    sampling_grid = vcat(x_t_flat, y_t_flat, all_ones)
+    sampling_grid = reshape(
+        transpose(repeat(transpose(sampling_grid), args[:bsz])),
+        3,
+        size(x_t_flat)[2],
+        args[:bsz],
+    )
+    return collect(Float32.(sampling_grid))
```

Comment:
```suggestion
    return Float32.(sampling_grid)
```
I don't think the cast to `Float32` should be necessary either, so you'll want to audit the rest of the function to make sure no funny type promotion is happening.